src/main/scala/zio/prelude/InverseNonZero.scala

```diff
@@ -0,0 +1,909 @@
+package zio.prelude
+
+import zio.prelude.coherent.EqualInverseNonZero
+import zio.test.TestResult
+import zio.test.laws.{ Lawful, Laws }
+
+/**
+ * The `InverseNonZero` type class describes an associative binary operator for a
+ * type `A` that has an identity element and an inverse binary operator.
+ * Combining any value with itself with the inverse operator must return the
+ * identity element. The exception may be the `zero` element, where the result is not defined.
+ * For example, for multiplication of reals, one is an identity
+ * element and division is an inverse operation, because dividing any
+ * value by itself always returns one.
+ *
+ * Because `InverseNonZero` defines a binary rather than a unary operator it can be
+ * used to describe inverse operations for types that do not have inverse
+ * values. For example, the natural numbers do not have inverses because the
+ * set of natural numbers does not include negative numbers. But we can still
+ * define a subtraction operation that is the inverse of addition for the
+ * natural numbers, since subtracting a number from itself always returns
+ * zero.
+ */
+trait InverseNonZero[A] extends Identity[A] {
```

Comment:
Unfortunately, I don't think this is very well defined, because the type class does not define a "zero" element. In general for an inverse, you don't need to speak about zero elements. 

We should think about other ways to achieve the goal of making this definable for division (or giving up). Such as `Inverse[A, B <: A]` or something.

Reply:
I have to admit that `InverseNonZero` is quite un-elegant and ugly :crying_cat_face:, I just didn't know how to do it better.

> Such as Inverse[A, B <: A] or something.

I'm not sure what you mean by this. Could you please expand on this? Hopefully this will get us to a better place :crossed_fingers: